docs/design/mscclpp-dsl-json-file.md

```diff
@@ -0,0 +1,561 @@
+
+Example:
+
+```json
+"channels": [
+  {
+    "type": "switch",
+    "buff": "i",
+    "rankGroups": [
+      {
+        "size": 8,
+        "ranks": [
+          0,
+          1,
+          2,
+          3,
+          4,
+          5,
+          6,
+          7
+        ]
+      }
+    ]
+  }
+]
+```
+
+### Remote Buffers
+The ```remoteBuffers``` field describes all the remote buffers that a given GPU needs to access. Each entry in this list contains the following fields:
+
+- ```rank``:  Indicates the rank that owns the remote buffer.
+- ```type```: Consist of the buffer type, this could have the following values: "i" for the input buffer, "o" for the output buffer, "s" for the scratch buffer.
+- ```infoLocation```: Depending on where the information is stored and where it needs to be accessed, the data may be transferred accordingly.
+
+Example
+
+```json
+"remoteBuffers": [
+  {
+    "rank": 1,
+    "type": "i",
+    "infoLocation": "gpu"
+  }
+]
+```
+
+### Thread Blocks
+The thread block field describe the operation inside each thread block, we have the following fields:
+
+- ```id```: The thread block id.
+- ```ops```: The list with all the operation in order they will be executed by this tread block.
+- ```channels```: The channels the thread block will use, referenced by the channel id. The channel id is based on the global channel description in the gpu, for example if the channel type is memory and the channel id is 0, it refers to the first channel of memory id type descriptioned in the gpu channels field.
+- ```remoteBuffersIds```: A list with all the remote buffer ids(related to the remove buffer field on the GPU) used by the thread block.
+
+For Example:
+
+```json
+"threadblocks": [
+  {
+    "id": 0,
+    "ops": [...],
+    "channels":[
+      {
+        "ctype": "memory",
+        "cids": [
+          0
+        ]
+      }
+    ],
+    "remoteBuffersIds": [
+      0
+    ]
+  }
+]
+```
+
+The most important field here is the ops fields, where there is the description for all the operations, each operation has specific fields, let's given an overview in all the possible fields and after this go through each of them:
+
+TO BE DEFINED
+
+#### Signal Operation
+The signal operation is composed by the field ```name```, ```cids```, ```ctype```.
+
+Example
+
+```json
+"ops": [
+  {
+    "name": "signal",
+    "cids": [
+      0
+    ],
+    "ctype": "memory",
+  }
+]
+```
+
+### Wait Operation
+The signal operation is composed by the field ```name```, ```cids```, ```ctype```.
+
+```json
+"ops": [
+  {
+    "name": "wait",
+    "cids": [
+      0
+    ],
+    "ctype": "memory",
+  }
+]
+```
+
+### Put Operation
+The put operation is composed by the field ```name```, ```src_buff```, ```dst_buff```, ```cids```, ```ctype```.
+
+```json
+"ops": [
+  {
+    "name": "put",
+    "src_buff": [
```

Comment:
As you mentioned before, how to deal with readReduceCopySend. It has remote_src_buff, local_dst_buff, remote_dst_buff. Also may need to move cid into buffer block, since we need to use this channel to send data. How about this:
```json
 {
              "name": "put",
              "local_src_buff": [
                {
                  "offset": 0,
                  "type": "i",
                }
              ],
              "remote_dst_buff": [
                {
                  "buff_id": 0,
                  "offset": 0,
                  "cid": 0,
                }
              ],
             "size": 1,
}
```